### PR TITLE
fix bug with sql stmt caching

### PR DESCRIPTION
**Pull request by adizaimi**
Statement caching is broken in several ways in R7. 
* Bug with refactored code -- behavior was slightly different compared to previous messy but correct code
* Need reset before putting vdbe stmt in cache
* Refactored more
* Added asserts for safety checks
* Remove from cache when in use and put back in cache when done so can't have two sp interleaving stmts use erroneously the same vdbe.

### CODE REVIEW

**Comment by akshatsikarwar** [comment]
Need to remove a bunch of debug trace.

**Comment by akshatsikarwar** [comment]
And resolve conflicts.

**Comment by akshatsikarwar** [comment]
Does this PR resolve issue: https://github.com/bloomberg/comdb2/issues/578 ?

**Comment by adizaimi** [comment]
Quite possibly. stmt caching should not be used or tested until this PR goes in. We might have other issues that this PR does not fix yet.

File: db/sql.h [hunk 1]
```diff
@@ -66,6 +66,7 @@ typedef struct stmt_hash_entry {
     struct schema *params_to_bind;
     struct stmt_hash_entry *prev;
```
Comment (akshatsikarwar) [inline db/sql.h]: why still have next, prev?

Comment (adizaimi) [reply db/sql.h]: thanks, good eyes.

File: tests/userschema.test/lrl.options [hunk 2]
```diff
@@ -1,3 +1,4 @@
 table   dummy   dummy.csc2
 
+enable_sql_stmt_caching none
```
Comment (akshatsikarwar) [inline tests/userschema.test/lrl.options]: remove

File: tests/setup [hunk 3]
```diff
@@ -117,6 +117,7 @@ if [[ -z $SKIPSSL ]] ; then
     cat >> $DBDIR/$DBNAME.lrl <<EOPTIONS
 ssl_client_mode REQUIRE
 ssl_cert_path $TESTDIR
+enable_sql_stmt_caching all
```
Comment (akshatsikarwar) [inline tests/setup]: remove from here once its made default

File: tests/sc_repeated_updates.test/lrl.options [hunk 4]
```diff
@@ -1,3 +1,4 @@
 table t2 t2.csc2
 round_robin_stripes
 ssl_client_mode OPTIONAL
+enable_sql_stmt_caching none
```
Comment (akshatsikarwar) [inline tests/sc_repeated_updates.test/lrl.options]: remove

File: tests/max_wr_rows.test/lrl.options [hunk 5]
```diff
@@ -0,0 +1 @@
+enable_sql_stmt_caching none
```
Comment (akshatsikarwar) [inline tests/max_wr_rows.test/lrl.options]: remove

File: tests/analyze.test/lrl.options [hunk 6]
```diff
@@ -1,3 +1,4 @@
 table t1 t1.csc2
 table t2 t2.csc2
 table t3 t3.csc2
+enable_sql_stmt_caching none
```
Comment (akshatsikarwar) [inline tests/analyze.test/lrl.options]: remove

File: tests/tunables.test/t00_all_tunables.expected [hunk 7]
```diff
@@ -284,7 +284,7 @@
 (name='enable_sock_fstsnd', description='', type='BOOLEAN', value='ON', read_only='Y')
 (name='enable_sp_strict_assignments', description='', type='INTEGER', value='0', read_only='Y')
 (name='enable_sparse_lockerid_map', description='If set, allocates a sparse map of lockers for deadlock resolution. (Default: on)', type='BOOLEAN', value='ON', read_only='Y')
-(name='enable_sql_stmt_caching', description='Enable caching of query plans. If followed by "all" will cache all queries, including those without parameters. (Default: off)', type='ENUM', value='NONE', read_only='Y')
+(name='enable_sql_stmt_caching', description='Enable caching of query plans. If followed by "all" will cache all queries, including those without parameters. (Default: on)', type='ENUM', value='ALL', read_only='Y')
```
Comment (akshatsikarwar) [inline tests/tunables.test/t00_all_tunables.expected]: Fix

**Comment by akshatsikarwar** [comment]
Minor cleanup reqd. Looks good otherwise. Thank you for fixing this!

**Pull request merged**